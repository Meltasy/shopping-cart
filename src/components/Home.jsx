import { NavLink, useParams } from 'react-router-dom'
import { useEffect, useState } from 'react'
import Shop from './Shop.jsx'
import Products from './Products.jsx'
import Cart from './Cart.jsx'

const Home = () => {
  const { name } = useParams()
  const [productItems, setProductItems] = useState(null)
  const [error, setError] = useState(null)
  const [loading, setLoading] = useState(true)
  // const [addToCart, setAddToCart] = useState([])
  // const [quantityCart, setQuantityCart] = useState(0)
  const [totalBasket, setTotalBasket] = useState(new Map())
  const [basketSize, setBasketSize] = useState(0)
  // const [buyProducts, setBuyProducts] = useState(null)

  useEffect(() => {
    fetch('https://fakestoreapi.com/products', { mode: 'cors' })
    .then((response) => {
      if (response.status >= 400) {
        throw new Error('Server error')
      }
      return response.json()
    })
    .then((response) => {
      console.log(response)
      setProductItems(response)
    })
    .catch((error) => setError(error))
    .finally(() => setLoading(false))
  }, [])

  if (loading) return <p>Loading...</p>
  if (error) return <p>Network error</p>

  function handleSubmit(itemId, value) {
    if (!totalBasket.has(itemId)) {
      setTotalBasket(totalBasket.set(itemId, value))
      console.log(totalBasket)
    } else {
      setTotalBasket(totalBasket.set(itemId, totalBasket.get(itemId) + value))
    }
    setBasketSize(totalBasket.size)
    // const products = productItems.map((item) => ({ ...item, quantity: 0 }))
    // setBuyProducts(products.map(item => {
    //   if (item.id === itemId) {
    //     console.log('Match', value, item.quantity)
    //     return {
    //       ...item,
    //       quantity: value
    //     }
    //   } else {
    //     return item
    //   }
    // }))
    // console.log(buyProducts)
    // setAddToCart(addToCart.map(item => {
    //   if (item.id === itemId) {
    //     return {...item, quantity: value}
    //   } else {
    //     console.log('inside addToCart.map', item, itemId)
    //   }
    // }))
    // setAddToCart([ ...addToCart, { id: itemId, quantity: value } ])
    // console.log('addToCart:',addToCart)
    // setQuantityCart(addToCart.length + 1)
    // console.log(addToCart.length)
  }

  return (
    <>
      <header>
        <h1>My shopping cart</h1>
        <p>What do you want to buy today?</p>
      </header>
      <nav>
        <ul>
          <li>
            <NavLink to='../shop'>Galeries Libellule</NavLink>
          </li>
          <li>
            <NavLink to='../products'>Atelier</NavLink>
          </li>
          <li>
            <NavLink to='../cart'>Panier</NavLink>
            {/* <div className='quantityCart'>{quantityCart}</div> */}
            <div className='quantityCart'>{basketSize}</div>
          </li>
        </ul>
      </nav>
      <main>
        {name === 'products' ? (
          <Products productItems={productItems} handleSubmit={handleSubmit}/>
        ) : name === 'cart' ? (
          // <Cart buyProducts={buyProducts} addToCart={addToCart}/>
          <Cart productItems={productItems} totalBasket={totalBasket}/>
        ) : (
          <Shop productItems={productItems}/>
        )}
      </main>
      <footer>
        <p>Meltasy's magnificent shopping mall</p>
      </footer>
    </>
  )
}

export default Home
