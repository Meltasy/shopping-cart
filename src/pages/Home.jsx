import { useParams } from 'react-router-dom'
import { useEffect, useState } from 'react'
import Navbar from '../components/Navbar.jsx'
import Intro from './Intro.jsx'
import Shop from './Shop.jsx'
import Cart from './Cart.jsx'

const Home = () => {
  const { name } = useParams()
  const [productItems, setProductItems] = useState(null)
  const [error, setError] = useState(null)
  const [loading, setLoading] = useState(true)
  const [addToCart, setAddToCart] = useState([])
  const [quantityCart, setQuantityCart] = useState(0)

  useEffect(() => {
    fetch('https://fakestoreapi.com/products', { mode: 'cors' })
    .then((response) => {
      if (response.status >= 400) {
        throw new Error('Server error')
      }
      return response.json()
    })
    .then((response) => {
      console.log(response)
      setProductItems(response)
    })
    .catch((error) => setError(error))
    .finally(() => setLoading(false))
  }, [])

  if (loading) return <p>Loading...</p>
  if (error) return <p>Network error</p>

  function handleSubmit(itemId, value) {
    setAddToCart([ ...addToCart, { id: itemId, quantity: value } ])
    console.log('addToCart:',addToCart)
    setQuantityCart(addToCart.length + 1)
    console.log(addToCart.length)
  }

  return (
    <>
      <Navbar quantityCart={quantityCart}/>
      <main>
        {name === 'shop' ? (
          <Shop productItems={productItems} handleSubmit={handleSubmit}/>
        ) : name === 'cart' ? (
          <Cart productItems={productItems} addToCart={addToCart}/>
        ) : (
          <Intro productItems={productItems}/>
        )}
      </main>
    </>
  )
}

export default Home
