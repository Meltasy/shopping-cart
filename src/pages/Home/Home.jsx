import { useState, useEffect } from 'react'
import styled from 'styled-components'

const Blurb = styled.p`
  font-size: 1.5rem;
  color: var(--primary-color);
  font-weight: 700;
  text-align: right;
  padding-right: 5rem;
  margin-bottom: 0;
`

const Wrapper = styled.div`
  display: flex;
  justify-content: space-around;
  gap: 20px;
  background-color: white;
  border-top: 20px solid var(--primary-color);
  border-bottom: 20px solid var(--primary-color);
  transform: skewY(10deg);
`

const StyledImage = styled.img`
  display: ${props => (props.$hidden ? 'none' : 'block' )};
  max-width: 500px;
  max-height: 500px;
  object-fit: contain;
  border-radius: 10px;
  padding: 20px;
`

const Home = ({ productItems }) => {
  const [currentImages, setCurrentImages] = useState([0, 1, 2])

  useEffect(() => {
    const interval = setInterval(() => {
      setCurrentImages(prevImages =>
        prevImages.map(image =>
          image === productItems.length - 1 ? 0 : image + 1
      ))
    }, 5000)
    return () => clearInterval(interval)
  }, [currentImages, productItems])

  return (
    <>
      <h1>Galeries Libellule</h1>
      <Blurb>See something you like?<br />
      Check it out in our shop!</Blurb>
      <Wrapper>
        {currentImages.map(index => {
          const item = productItems[index]
          return (
            item && (
              <StyledImage
                key={item.id}
                src={item.image}
                alt={item.title}
                $hidden={false}
              />
            )
          )
        })}
      </Wrapper>
    </>
  )
}

export default Home
