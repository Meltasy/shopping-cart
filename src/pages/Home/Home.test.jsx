import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react'
import Home from './Home'

const mockProductItems = [
  { id: '123', image: 'fakeImageA.jpg', title: 'Fake product A' },
  { id: '321', image: 'fakeImageB.jpg', title: 'Fake product B' },
  { id: '231', image: 'fakeImageC.jpg', title: 'Fake product C' },
]

describe('Home component', () => {
  it('renders galeries libellule', () => {
    const { container } = render(<Home productItems={mockProductItems}/>)
    expect(container).toMatchSnapshot()
  })

  it ('renders correct heading', () => {
    render(<Home productItems={mockProductItems}/>)
    screen.debug()
    expect(screen.getByRole('heading').textContent).toMatch(/galeries libellule/i)
    expect(screen.getByRole('heading', { name: 'Galeries Libellule' })).toBeInTheDocument()
  })

  it('renders home products with all complete details', () => {
    render(<Home productItems={mockProductItems} />)
    expect(screen.getAllByRole('img')[2]).toHaveAttribute('src', 'fakeImageC.jpg')
    expect(screen.getAllByAltText('Fake product C')[2])
  })

  it('renders correct number of home products', () => {
    render(<Home productItems={mockProductItems} />)
    const productItems = screen.getAllByRole('img')
    expect(productItems).toHaveLength(mockProductItems.length)
  })
})
