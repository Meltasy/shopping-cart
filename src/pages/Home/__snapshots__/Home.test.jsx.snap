// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home component > renders galeries libellule 1`] = `
<div>
  <h1>
    Galeries Libellule
  </h1>
  <p
    class="sc-FEMpB gTvDhS"
  >
    See something you like?
    <br />
    Check it out in our shop!
  </p>
  <div
    class="sc-gjcoXW kBAqGb"
  >
    <img
      alt="Fake product A"
      class="sc-dVBluf ideMSn"
      src="fakeImageA.jpg"
    />
    <img
      alt="Fake product B"
      class="sc-dVBluf ideMSn"
      src="fakeImageB.jpg"
    />
    <img
      alt="Fake product C"
      class="sc-dVBluf ideMSn"
      src="fakeImageC.jpg"
    />
  </div>
</div>
`;
